File: js/base/SFLord.js
;
/*!
 *
 *  SpriteForest JavaScript Library v1.0.0
 *  http://github.com/moky/SpriteForest/js/
 *
 *  Copyright (c) 2014 moKy at slanissue.com
 *  Released under the MIT license
 *  http://en.wikipedia.org/wiki/MIT_License
 *
 */
//
//  SFLord.js
//  SpriteForest-HTML5
//
//  Created by moKy on Mar. 1, 2014.
//

!function(sf) {
	'use strict';
	var cc = sf.cc;
	var cn = sf.cn;
	
	//
	//  constructor
	//
	function Lord() {
		this._currentForest = null;
		this._histories = null;
	}
	
	//
	//  static functions
	//
	var _sharedLoard = null;
	
	function getInstance() {
		if (!_sharedLoard) {
			_sharedLoard = new sf.Lord();
			_sharedLoard.init();
		}
		return _sharedLoard;
	}
	
	function isForestExists() {
		//...
		return true;
	}
	
	//
	//  instance functions
	//
	function init() {
		this._currentForest = null;
		this._histories = [];
		return true;
	}
	
	function runIntoForest(path, scene) {
		
		if (!sf.Lord.isForestExists(path)) {
			cn.log('forest not exists at path: ' + path);
			return false;
		}
		
		var his = {};
		his['File'] = path;
		this._histories.push(his);
		
		cn.log('running into forest: ' + path);
		this._currentForest = sf.Forest.create(path);
		return this.performScene(scene);
	}
	
	function performScene(scene) {
		if (!this._currentForest) {
			cn.error(scene);
			return false;
		}
		
		var obj = null;
		if (!scene) {
		} else if (scene instanceof cc.Scene) {
			obj = scene;
		} else if (typeof scene === 'object') {
			obj = scene['Scene'];
		}
		var his = {};
		his['Scene'] = obj;
		this._histories.push(his);
		
		return this._currentForest.performScene(scene);
	}
	
	//////////////////////////////////
	//
	//  history stack:
	//
	//        (0)   File : "/path/to/forest/root/default.bundle/main.plist"
	//        (1)   Scene: [NSNull null]
	//        (2)   Scene: "scene2"
	//        (x)   ...
	//        (x+1) File : "/path/to/forest/root/forest2.bundle/main.plist"
	//        (x+2) Scene: [NSNull null]
	//
	function goBack() {
		// lock begin
		
		var histories = this._histories;
		
		if (histories.length <= 2) {
			cn.log('reach the first scene');
			return false;
		}
		
		histories.pop(); // remove last object
		
		var last = histories[histories.length - 1];
		
		var path;
		if (last['File']) {
			// it's the bounds of anthor forest
			histories.pop();
			
			// search the previous forest's path
			for (var i = histories.length - 1; i >= 0; --i) {
				var dict = histories[i];
				path = dict['File'];
				if (path) {
					cn.log('path = ' + path);
					break;
				}
			}
		}
		
		last = histories[histories.length - 1];
		var data = last ? last['Scene'] : null;
		
		histories.pop();
		
		// do it!
		if (path) {
			this.runIntoForest(path, data);
			if (histories.length > 2) {
				// remove duplicated 'File' record
				histories.splice(histories.length - 2, 1);
			}
		} else {
			this.performScene(data);
		}
		
		// lock end
		return true;
	}
	
	function showLoadingIndicator() {
		// do nothing
	}
	
	function hideLoadingIndicator() {
		// do nothing
	}
	
	//--------------------------------------------------------------------------
	
	var prop = {
		ctor: Lord,
		
		init: init,
		
		runIntoForest: runIntoForest,
		performScene: performScene,
		
		goBack: goBack,
		
		// for loading indicator
		showLoadingIndicator: showLoadingIndicator,
		hideLoadingIndicator: hideLoadingIndicator,
		
		// properties
		_currentForest: null,
		_histories: [],
	};
	
	// extends
	sf.Lord = cc.Class.extend(prop);
	
	// static
	sf.Lord.getInstance = getInstance;
	
	sf.Lord.isForestExists = isForestExists;
	
}(SpriteForest);

File: x/base/SFLord.cpp
//
//  SFLord.cpp
//  SpriteForest-X
//
//  Created by Moky on 12-12-4.
//  Copyright 2012 Slanissue Inc. All rights reserved.
//

#include "SFLord.h"

NS_SF_BEGIN

USING_NS_CC;
USING_NS_CN;

SFLord::SFLord(void)
: CCObject()
, m_pCurrentForest(NULL)
, m_pHistory(NULL)
{
	
}

SFLord::~SFLord(void)
{
	setCurrentForest(NULL);
	setHistory(NULL);
}

bool SFLord::init(void)
{
	setCurrentForest(NULL);
	setHistory(CCArray::arrayWithCapacity(29));
	return true;
}

static SFLord * m_pSharedLord = NULL;

SFLord * SFLord::sharedLord(void)
{
	if (!m_pSharedLord)
	{
		m_pSharedLord = new SFLord();
		if (m_pSharedLord && m_pSharedLord->init())
		{
			CNLog("new lord born!");
		}
		else
		{
			CNLog("could not happen!");
			CC_SAFE_DELETE(m_pSharedLord);
			m_pSharedLord = NULL;
		}
	}
	return m_pSharedLord;
}

void SFLord::end(void)
{
	if (m_pSharedLord)
	{
		CNLog("my lord is abdicating...");
	}
	CC_SAFE_RELEASE_NULL(m_pSharedLord);
}

bool SFLord::isForestExists(const char * pszPath)
{
	if (!pszPath || strlen(pszPath) == 0)
	{
		CNLog("error");
		return false;
	}
	
	std::string path(pszPath);
	std::string ext(pszPath);
	ext = CNString::pathExtension(ext);
	if (strcasecmp(ext.c_str(), "plist") != 0)
	{
		path = CNString::stringByAppendingPathComponent(path, "main.plist");
	}
	
	return CNFileManager::isExists(path.c_str());
}

bool SFLord::runIntoForest(const char * pszPath, CCObject * scene = NULL)
{
	if (!pszPath || strlen(pszPath) == 0)
	{
		CNLog("invalid path");
		return false;
	}
	std::string str(pszPath);
	{
		std::size_t pos1, pos2;
		while ((pos1 = str.find("/../")) != std::string::npos)
		{
			pos2 = str.rfind("/", pos1 - 1);
			if (pos2 == std::string::npos)
			{
				pos2 = - 1;
			}
			str = str.substr(0, pos2 + 1) + str.substr(pos1 + 4);
		}
		CNLog("convert string '%s' to '%s'", pszPath, str.c_str());
		pszPath = str.c_str();
	}
	
	if (!isForestExists(pszPath))
	{
		CNLog("forest not exists at path: %s", pszPath);
		return false;
	}
	
	CCDictionary * pHis = CCDictionary::dictionary();
	pHis->setObject(CCString::stringWithCString(pszPath), "File");
	m_pHistory->addObject(pHis);
	
	CNLog("running into forest: %s", pszPath);
	setCurrentForest(SFForest::forest(pszPath));
	return performScene(scene);
}

bool SFLord::performScene(CCObject * scene) const
{
	if (!m_pCurrentForest)
	{
		CNLog("error");
		return false;
	}
	
	CCObject * pObj = NULL;
	if (CCDictionary * pDict = dynamic_cast<CCDictionary *>(scene))
	{
		pObj = pDict->objectForKey("Scene");
	}
	else if (scene)
	{
		pObj = scene;
	}
	if (!pObj)
	{
		pObj = new CCObject();
		pObj->autorelease();
	}
	CCDictionary * pHis = CCDictionary::dictionary();
	pHis->setObject(pObj, "Scene");
	m_pHistory->addObject(pHis);
	
	return m_pCurrentForest->performScene(scene);
}

//////////////////////////////////
//
//  history stack:
//
//        (0)   File : "/path/to/forest/root/default.bundle/main.plist"
//        (1)   Scene: [NSNull null]
//        (2)   Scene: "scene2"
//        (x)   ...
//        (x+1) File : "/path/to/forest/root/forest2.bundle/main.plist"
//        (x+2) Scene: [NSNull null]
//
bool SFLord::goBack(void)
{
	// lock begin
	
	if (!m_pHistory)
	{
		CNLog("could not happen");
		return false;
	}
	
	if (m_pHistory->count() <= 2)
	{
		CNLog("reach the first scene");
		return false;
	}
	
	m_pHistory->removeLastObject();
	
	CCDictionary * pLast = (CCDictionary *)m_pHistory->lastObject();
	if (!pLast)
	{
		CNLog("could not happen");
		m_pHistory->removeLastObject();
		return false;
	}
	
	CCString * pPath = NULL;
	if (pLast->objectForKey("File"))
	{
		// it's the bounds of anthor forest
		m_pHistory->removeLastObject();
		
		// searchig the previous forest's path
		for (int pos = m_pHistory->count() - 1; pos >= 0; pos--)
		{
			CCDictionary * pDict = (CCDictionary *)m_pHistory->objectAtIndex(pos);
			if (!pDict)
			{
				CNLog("could not happen");
				continue;
			}
			pPath = (CCString *)pDict->objectForKey("File");
			if (pPath)
			{
				CNLog("path = %s", pPath->getCString());
			}
		}
	}
	
	CCObject * pData = NULL;
	pLast = (CCDictionary *)m_pHistory->lastObject();
	if (pLast)
	{
		pData = pLast->objectForKey("Scene");
	}
	
	CC_SAFE_RETAIN(pPath);
	CC_SAFE_RETAIN(pData);
	m_pHistory->removeLastObject();
	
	// do it!
	if (pPath)
	{
		runIntoForest(pPath->getCString(), pData);
		if (m_pHistory->count() > 2)
		{
			// remove duplicated 'File' record
			m_pHistory->removeObjectAtIndex(m_pHistory->count() - 2);
		}
	}
	else
	{
		performScene(pData);
	}
	
	CC_SAFE_RELEASE(pPath);
	CC_SAFE_RELEASE(pData);
	
	// lock end
	
	return true;
}

//void SFLord::showLoadingIndicator(void)
//{
//	CNLog("not implement");
//}
//
//void SFLord::hideLoadingIndicator(void)
//{
//	CNLog("not implement");
//}

NS_SF_END
